examples/with-typescript-graphql/lib/with-apollo.tsx

```diff
@@ -54,11 +73,25 @@ export default function withApollo(
 
   if (ssr || PageComponent.getInitialProps) {
     WithApollo.getInitialProps = async (ctx: WithApolloPageContext) => {
-      const { AppTree } = ctx
+      // Resolver context here is only set on server. For client-side,
+      // "/api/graphql" route creates and pass it to resolver functions.
+      let resolverContext: ResolverContext | undefined
+      // Keep the "isServer" check inline, so webpack removes the block
+      // for client-side bundle.
+      if (typeof window === 'undefined') {
+        if (!ctx.req || !ctx.res) throw new Error('never')
+        resolverContext = await createResolverContext({
+          req: ctx.req,
+          res: ctx.res,
+        })
+      }
```

Comment:
What would be the use case for this?

Reply:
Thank you for asking. It's necessary for SSR to pass a resolver context to resolver functions. For SSR, it:

0. Starts from `WithApollo.getInitialProps`
1. Creates a resolver context from `req` and `res`
2. Initializes an apollo client instance with [`SchemaLink`](https://github.com/zeit/next.js/pull/11681/files#diff-52c0088e3df4611902ee3b5f59a4a97dR196) with the resolver context
3. Executes `getDataFromTree`, which leads the apollo client to...
4. Pass the resolver context to [resolver functions as the 3rd argument](https://github.com/zeit/next.js/pull/11681/files#diff-682d12e5f9468d1190b78c5c58a34533R5)

To pass a resolver context to resolver functions, we have to take care of two paths.

1. One for an  HTTP request to the GraphQL server (it's in `pages/api/graphql.ts`)
1. Another for SSR, done by `SchemaLink` that allows us to resolve GraphQL query without calling HTTP request.

The above code covers the case no. 2.

As you know, both paths run on the server-side, so here we have to get off client-side execution.

Reply:
@piglovesyou Thank you, it's very clear now! :relieved: